Fix compatibility issues with QT Creator versions 15.0.1 and 16.0.0

PolymorphicAgent opened the pull request:
Fixes compatibility issues with QT Creator 15.0.1.

PolymorphicAgent committed:
Update cutediscordpresence.h

Fixes compatibility issues with QT Creator 15.0.1

PolymorphicAgent committed:
Update build_cmake.yml

Move from deprecated 'upload-artifact@v2' to 'upload-artifact@v3'.

eduardoc7 commented on .github/workflows/build_cmake.yml:
```diff
@@ -54,7 +54,7 @@ jobs:
 
       - name: Upload release url
-        uses: actions/upload-artifact@v2
+        uses: actions/upload-artifact@v3
```
```suggestion
        uses: actions/upload-artifact@v4
```

eduardoc7 commented on .github/workflows/build_cmake.yml:
```diff
@@ -408,15 +408,15 @@ jobs:
         Write-Output "PLUGIN_ARCHIVE_NAME=${env:PLUGIN_ARCHIVE_NAME}" | Out-File -FilePath $env:GITHUB_ENV -Encoding utf8 -Append
 
-    - uses: actions/upload-artifact@v2
+    - uses: actions/upload-artifact@v3
```
```suggestion
    - uses: actions/upload-artifact@v4
```

eduardoc7 commented on .github/workflows/build_cmake.yml:
```diff
@@ -408,15 +408,15 @@ jobs:
     - name: Download release url
       if: contains(github.ref, '/tags/v')
-      uses: actions/download-artifact@v2
+      uses: actions/download-artifact@v3
```
```suggestion
      uses: actions/download-artifact@v4
```

PolymorphicAgent committed:
Update cutediscordpresence.cpp

Fix the error '"Id" is missing' when loading plugin

eduardoc7 commented on cutediscordpresence.h:
```diff
@@ -30,10 +30,10 @@
-constexpr const char* GLOBAL_DRPC_CONTROL_MENU_START_ACTION_ID    { "QtCreatorDRPC.Action.Start" };
-constexpr const char* GLOBAL_DRPC_CONTROL_MENU_STOP_ACTION_ID     { "QtCreatorDRPC.Action.Stop" };
-constexpr const char* GLOBAL_DISCORD_APPLICATION_ID               { "937400240473006092" };
+#define GLOBAL_DRPC_CONTROL_MENU_ID               "QtCreatorDRPC.Menu"
+#define GLOBAL_DRPC_CONTROL_MENU_START_ACTION_ID  "QtCreatorDRPC.Action.Start"
+#define GLOBAL_DRPC_CONTROL_MENU_STOP_ACTION_ID   "QtCreatorDRPC.Action.Stop"
+#define GLOBAL_DISCORD_APPLICATION_ID             "937400240473006092"
```
Why is this change necessary?

Using constexpr is generally preferred over #define because it provides type safety, respects scope, and allows for better debugging

PolymorphicAgent replied:
I'm attempting to fix errors related to "Id" is missing when the plugin gets loaded - I thought that might help but aparently I'm wrong (the error is still occuring)

![image](https://github.com/user-attachments/assets/581d246f-59f7-44af-ab4d-73be915ece3b)

eduardoc7 replied:
Alright, in this case, we can revert the change I mentioned earlier.

Regarding the issue you're facing, it might be related to an update in the plugin loading process. I checked the documentation, and now the ID field is required in the json.in file of the project... Could you please add this new field there and verify?
Ref: https://doc.qt.io/qtcreator-extending/first-plugin.html#plugin-meta-data-template

PolymorphicAgent replied:
sure! (pretty busy with school so It might be a bit)

eduardoc7 replied:
No problem, this afternoon I'll create a branch and fix this Qt Creator compatibility issue to help you.

eduardoc7 reviewed (changes requested):
Suggestion to use the latest version of the GitHub Actions artifacts and a question regarding the use of macros.

Thanks for your contribution! 😉

PolymorphicAgent committed:
Update cutediscordpresence.cpp

Original commit was a mistake. setID doesn't exist.

PolymorphicAgent committed:
Update CuteDiscordPresence.json.in

Fixing missing "ID" field runtime error

PolymorphicAgent committed:
Update CuteDiscordPresence.json.in

Plugin id "Cute Discord Presence" must be lowercase.

PolymorphicAgent committed:
Update CuteDiscordPresence.json.in

Added "VendorId"

PolymorphicAgent committed:
Update CuteDiscordPresence.json.in

forgot a ',' (oops!)

PolymorphicAgent committed:
Update CMakeLists.txt

attempt to explicitly define plugin dependency id's. Not sure if this will resolve the runtime error: Dependency: "Id" is missing.

PolymorphicAgent committed:
Update CMakeLists.txt

revert previous commit: did not work

PolymorphicAgent committed:
Update CuteDiscordPresence.json.in

Attempt to override automatic moc's changes: it wasn't properly including the dependency ids.

PolymorphicAgent committed:
Update CuteDiscordPresence.json.in

Fix dependency versions (hopefully?)

PolymorphicAgent committed:
Update CuteDiscordPresence.json.in

Fix Discord-rpc version (maybe?)

PolymorphicAgent commented:
I've gotten it to a point where I just need the correct versions for all of the dependencies - automoc wasn't including the "ID" and "VendorID" for each dependency when processing the .json.in, so I think (?) the best thing to do is just override automoc and manually put in the dependencies? unless automoc can be correctly configured in the cmakelist?

PolymorphicAgent committed:
Update CuteDiscordPresence.json.in

attempt to fix versions (again)

PolymorphicAgent committed:
Update build_cmake.yml

Started complaining about artifact v3?!?!?

PolymorphicAgent committed:
Update build_cmake.yml

update download-artifact version to v4

PolymorphicAgent committed:
Update build_cmake.yml

fix upload-artifacts version (v4)

PolymorphicAgent committed:
Update CuteDiscordPresence.json.in

This might work now that I updated the qt creator version in the workflow?

PolymorphicAgent commented:
Alright! Everything works and tested with the latest qt creator (15.0.1). I just re-added the '${IDE_PLUGIN_DEPENDENCIES}' to the .json.in file and updated the qt creator version in the .yaml build configuration file.

PolymorphicAgent commented:
oops

eduardoc7 commented:
> Alright! Everything works and tested with the latest qt creator (15.0.1). I just re-added the '${IDE_PLUGIN_DEPENDENCIES}' to the .json.in file and updated the qt creator version in the .yaml build configuration file.

Alright, I got a bit lost with the changes here in the commits... I have a question: on which platform did you test it, and did it work? Also, please revert this change in the defines so that it doesn't get pushed https://github.com/eduardoc7/qtcreator-plugin-discord-presence/pull/2#discussion_r1934494314

PolymorphicAgent committed:
Update cutediscordpresence.h

Revert changes

PolymorphicAgent commented:
Reverted the requested changes.
I tested it on Windows x64 with qt creator verison 15.0.1, and it works. Did not test on linux.

PolymorphicAgent commented:
ah, now I remember why I changed them to defines. It won't compile with constexpr: [Workflow Run Failure](https://github.com/PolymorphicAgent/qtcreator-plugin-discord-presence/actions/runs/13081212878/job/36504955711#step:7:167)

PolymorphicAgent committed:
Update cutediscordpresence.h

I know this isn't proper, but the project will not compile without these as defines.

PolymorphicAgent committed:
Update build_cmake.yml

Update to qt creator version 16.0.0, and qt version 6.8.2

PolymorphicAgent committed:
Update build_cmake.yml

update to msvc 2022

PolymorphicAgent committed:
Update build_cmake.yml

Major changes; copied sections from qt-creator-16.0.0's generated build_cmake.yml and commented others. Hopefully this works?

PolymorphicAgent committed:
Update build_cmake.yml

cut unnecessary platforms from build

PolymorphicAgent commented:
I updated the workflow file to be compatible with qt 6.8.2 and qt creator 16.0.0 (Admittedly, I don't know if I broke the release flow) and the main header file `CuteDiscordPresence.h` back to using `#define` (It wouldn't compile otherwise). I tested the plugin successfully on the windows platform.

QT Creator version info:
```
Product: Qt Creator 16.0.0
Based on: Qt 6.8.2 (MSVC 2022, x86_64)
Built on: Mar 11 2025 11:04:42
From revision: 55ce4e9c06
```

eduardoc7 commented:
> I updated the workflow file to be compatible with qt 6.8.2 and qt creator 16.0.0 (Admittedly, I don't know if I broke the release flow) and the main header file `CuteDiscordPresence.h` back to using `#define` (It wouldn't compile otherwise). I tested the plugin successfully on the windows platform.
> 
> QT Creator version info:
> 
> ```
> Product: Qt Creator 16.0.0
> Based on: Qt 6.8.2 (MSVC 2022, x86_64)
> Built on: Mar 11 2025 11:04:42
> From revision: 55ce4e9c06
> ```

How did you test this without generating the release artifact? This needs to be working for us to accept this PR...
I'm a bit busy at the moment, so if you can sort this out, I'll test it on Linux to check if the changes are okay.

PolymorphicAgent commented:
I just grabbed the artifact from the [artifacts section](https://github.com/PolymorphicAgent/qtcreator-plugin-discord-presence/actions/runs/13946305634) of the workflow run (as you can probably tell, I'm really new to anything git and don't know how to properly use it...)

I'll try my best to figure out how to generate the release artifact!

eduardoc7 commented:
> I just grabbed the artifact from the [artifacts section](https://github.com/PolymorphicAgent/qtcreator-plugin-discord-presence/actions/runs/13946305634) of the workflow run (as you can probably tell, I'm really new to anything git and don't know how to properly use it...)
> 
> I'll try my best to figure out how to generate the release artifact!

I'm currently running some tests to check the compatibility of QtCreator 15 on this branch and in this workflow: https://github.com/eduardoc7/qtcreator-plugin-discord-presence/actions/runs/13949599024. Which version of Qt Creator are you trying to use? Could you test the generated artifact for Windows? @PolymorphicAgent

PolymorphicAgent committed:
Update build_cmake.yml

Fix release workflow: Attempt #1

PolymorphicAgent committed:
Finalize workflow release logic

PolymorphicAgent commented:
Sorry for the delayed response!

I'm working with qt creator 16.0.0, and I tested [these](https://github.com/PolymorphicAgent/qtcreator-plugin-discord-presence/actions/runs/13953330210) generated artifacts for windows. The test was successful as the plugin loaded and functioned properly.

I also [fixed the release job](https://github.com/PolymorphicAgent/qtcreator-plugin-discord-presence/commit/548f9eec7608b591f5dba1a963177fa7e92fa6d7), and tested that to verify that it [creates a release properly](https://github.com/PolymorphicAgent/qtcreator-plugin-discord-presence/releases/tag/v1.0.3-Qt-6.8.2).

I will test the artifacts from the run you mentioned here

> ...in this workflow: https://github.com/eduardoc7/qtcreator-plugin-discord-presence/actions/runs/13949599024. Which version of Qt Creator are you trying to use? Could you test the generated artifact for Windows? 

on windows for qt creator 15 when I get home from school.

eduardoc7 commented:
I made some small modifications to achieve compatibility with QTC 15 and QTC 16, and it worked on both versions on Ubuntu 22.04. Please, when you have the chance, test this [latest release](https://github.com/eduardoc7/qtcreator-plugin-discord-presence/releases/tag/v0.2.0) I generated, which fixes this compatibility issue, so we can mark it as resolved and close this PR

If you can, leave a star on the repository! 😊
@PolymorphicAgent

PolymorphicAgent commented:
I was able to test for QTC 16 (and it worked as expected), but when I went to install QTC 15 from the maintenance tool, It wasn't under any menu (even after enabling historical versions). Is it really worth it for me to look for another way to install QTC 15 just to test this plugin?

eduardoc7 commented:
> I was able to test for QTC 16 (and it worked as expected), but when I went to install QTC 15 from the maintenance tool, It wasn't under any menu (even after enabling historical versions). Is it really worth it for me to look for another way to install QTC 15 just to test this plugin?

If this worked on QTC 16 and that's the version you use, there's no need to test on 15. Just have fun!

PolymorphicAgent commented:
Thanks, and starred the repo! Maybe by the time that QTC 17 comes out I'll know how to properly make a pr 💀
